overlord/state/taskrunner.go

```diff
@@ -78,16 +78,18 @@ func taskFail(t *Task) {
 // run must be called with the state lock in place
 func (r *TaskRunner) run(fn HandlerFunc, task *Task) {
 	task.SetStatus(RunningStatus) // could have been set to waiting
-	tomb := &tomb.Tomb{}
-	r.tombs[task.ID()] = tomb
-	tomb.Go(func() error {
-		err := fn(task, tomb)
+	tb := &tomb.Tomb{}
+	r.tombs[task.ID()] = tb
+	tb.Go(func() error {
+		err := fn(task, tb)
 
 		r.state.Lock()
 		defer r.state.Unlock()
 		halted := task.HaltTasks()
-		if err == nil {
+		if err == nil && tb.Err() == tomb.ErrStillAlive {
```

Comment:
That doesn't feel right. If something asks the tomb to stop for whatever reason, it won't return that, but it doesn't mean the task isn't done. What's the actual goal here?